In [1]:
import pandas as pd
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\duong\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [2]:
df = pd.read_csv('Data Scientist.csv')
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 977 entries, 0 to 976
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   title            977 non-null    object
 1   place            977 non-null    object
 2   company          977 non-null    object
 3   posting_date     977 non-null    object
 4   link             977 non-null    object
 5   job_description  977 non-null    object
 6   jd_html          977 non-null    object
 7   seniority_level  977 non-null    object
 8   job_function     977 non-null    object
 9   employment_type  977 non-null    object
 10  industries       977 non-null    object
dtypes: object(11)
memory usage: 84.1+ KB


title                         place  \
0             Machine Learning & Data Engineer                     Plano, TX   
1  Machine Learning Engineer — Computer Vision                Richardson, TX   
2       Software and Machine Learning Engineer                 Cambridge, MA   
3                    Machine Learning Engineer  Greater Syracuse-Auburn Area   
4                    Machine Learning Engineer                 Hillsboro, OR   

                                 company posting_date  \
0                   JPMorgan Chase & Co.   2021-02-22   
1                              Honeywell   2020-12-21   
2  Massachusetts Institute of Technology   2021-01-08   
3                                 Oracle   2021-01-21   
4                      Intel Corporation   2020-10-15   

                                                link  \
0  https://www.linkedin.com/jobs/view/machine-lea...   
1  https://www.linkedin.com/jobs/view/machine-lea...   
2  https://www.linkedin.com/jobs/view/software-an...   
3  https://www.linkedin.com/jobs/view/machine-lea...   
4  https://www.linkedin.com/jobs/view/machine-lea...   

                                     job_description  \
0  Corporate Banking Technology is hiring a multi...   
1  Corporate Banking Technology is hiring a multi...   
2  Join a team recognized for leadership, innovat...   
3  \nWorking at MIT offers opportunities, an envi...   
4  Design, develop, troubleshoot and debug softwa...   

                                             jd_html seniority_level  \
0  <div class="description__text description__tex...     Entry level   
1  <div class="description__text description__tex...     Entry level   
2  <div class="description__text description__tex...     Entry level   
3  <div class="description__text description__tex...     Entry level   
4  <div class="description__text description__tex...     Entry level   

                          job_function employment_type  \
0  Engineering, Information Technology       Full-time   
1  Engineering, Information Technology       Full-time   
2  Engineering, Information Technology       Full-time   
3  Engineering, Information Technology       Full-time   
4  Engineering, Information Technology       Full-time   

                                          industries  
0                                            Banking  
1                                            Banking  
2  Staffing and Recruiting, Human Resources, Acco...  
3  Nonprofit Organization Management, Internet, I...  
4  Information Technology and Services, Internet,...

## 1. Duplication in job description

In [5]:
df.job_description.duplicated().sum()

186

There are 186 dupicated description among 977 entries. For some reasons, the scraper packages retreived incorrect description
My task is now correct the link between job and job description

In [14]:
# For example, description #2 is belong to the job # 1
print(df.job_description[2])

Join a team recognized for leadership, innovation and diversity

Honeywell is a Fortune 100 company with global sales surpassing $40B and has been one of Fortune’s Most Admired Companies for over a decade. Through innovation, the Company brings together the physical and digital world to tackle some of the toughest societal and business problems – making the world a more productive, safe, and sustainable place.

We’re looking for a new senior team member who is motivated by cracking tough challenges with data, trained in problem solving, and with an unending thirst for learning.

Have endless opportunities to progress your career as you develop new offerings that impact a wide range of industries, improve the quality of life, and truly change the world.

As a Machine Learning Engineer with Honeywell, this is your opportunity to develop machine learning models for computer vision applications to address unique and complex problems for the variety of breakthrough solutions.

Responsibilit

In [8]:
df.job_description[0].find(df.job_description[0])

0

In [81]:
# Lower all strings
df['job_description'] = df.job_description.str.lower()
df.dropna(inplace=True)
df.head(3)

title            place       company posting_date  \
0  Data Engineer  Los Angeles, CA     Snap Inc.   2021-02-19   
1  Data Engineer     New York, NY  DigitalOcean   2021-02-18   
2  Data Engineer       Remote, OR     Roofstock   2021-02-18   

                                                link  \
0  https://www.linkedin.com/jobs/view/data-engine...   
1  https://www.linkedin.com/jobs/view/data-engine...   
2  https://www.linkedin.com/jobs/view/data-engine...   

                                     job_description  \
0  snap inc. is a camera company. we believe that...   
1  have you ever wondered what happens inside the...   
2  roofstock is the leading marketplace for inves...   

                                             jd_html seniority_level  \
0  <div class="description__text description__tex...     Entry level   
1  <div class="description__text description__tex...     Entry level   
2  <div class="description__text description__tex...     Entry level   

                          job_function employment_type  \
0               Information Technology       Full-time   
1               Information Technology       Full-time   
2  Information Technology, Engineering       Full-time   

                                          industries  
0  Marketing and Advertising, Computer Software, ...  
1  Marketing and Advertising, Computer Software, ...  
2  Information Technology and Services, Computer ...

In [82]:
print(df.job_description[2])

roofstock is the leading marketplace for investing in single-family rental homes. our mission is to make real estate investing radically accessible, cost effective and simple. we want to use technology to transform the way real estate is bought and sold and make real estate investing as simple as investing in stocks. simply put, we are passionate about helping our customers build wealth through real estate.

roofstock has been recognized as a great workplace great place to work® and have been recently named to the forbes fintech 50 and the red herring 100 lists of most innovative companies. roofstock has raised over $135m to date, is based in oakland, ca and dallas,tx with approximately 180 people and is growing rapidly. check out our reviews and see why our employees love working here!

we've built an amazing culture where everyone has a voice. we're nice. we get things done. we have fun. we believe in work-life balance (yes, you can really take a vacation). and our values are pretty 

tokenize → remove stopword, special characters → Stemming → Filter → Find common unigrams, bigrams, trigrams

Frequency, PCA or SVD, Fourier TF?, classification entry vs intern, neural network

In [83]:
df.dropna(inplace=True)

In [84]:
from nltk.tokenize import RegexpTokenizer

tokenizer = RegexpTokenizer(r'\w+')
tokenizer.tokenize('Eighty-seven miles to go, yet.  Onward!')

['Eighty', 'seven', 'miles', 'to', 'go', 'yet', 'Onward']

In [85]:
# word tokenize
df['job_description'] = df.job_description.apply(tokenizer.tokenize)
df.job_description

0      [snap, inc, is, a, camera, company, we, believ...
1      [have, you, ever, wondered, what, happens, ins...
2      [roofstock, is, the, leading, marketplace, for...
3      [the, data, engineer, is, responsible, for, pr...
4      [locations, tx, plano, united, states, of, ame...
                             ...                        
321    [feedback, loop, feedbackloop, com, is, a, gro...
322    [do, you, believe, that, creators, should, hav...
323    [embark, trucks, is, a, leading, self, driving...
324    [why, choose, between, doing, meaningful, work...
325    [about, may, mobility, may, mobility, is, a, s...
Name: job_description, Length: 326, dtype: object

In [65]:
# remove stopwords
from nltk.corpus import stopwords
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\duong\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [86]:
all_stopwords = stopwords.words('english')
all_stopwords[:2]

['i', 'me']

In [93]:
df['job_description'] = df.job_description.apply(lambda text: [w for w in text if w not in all_stopwords])
df.job_description

0      [snap, inc, camera, company, believe, reinvent...
1      [ever, wondered, happens, inside, cloud, based...
2      [roofstock, leading, marketplace, investing, s...
3      [data, engineer, responsible, processing, stru...
4      [locations, tx, plano, united, states, america...
                             ...                        
321    [feedback, loop, feedbackloop, com, growth, st...
322    [believe, creators, ability, get, paid, value,...
323    [embark, trucks, leading, self, driving, truck...
324    [choose, meaningful, work, fulfilling, life, m...
325    [may, mobility, may, mobility, self, driving, ...
Name: job_description, Length: 326, dtype: object

In [101]:
porter = nltk.PorterStemmer()
print(df.job_description[0][:20])
print([porter.stem(t) for t in df.job_description[0][:20]])
wnl = nltk.WordNetLemmatizer()
print([wnl.lemmatize(t) for t in df.job_description[0][:20]])

['snap', 'inc', 'camera', 'company', 'believe', 'reinventing', 'camera', 'represents', 'greatest', 'opportunity', 'improve', 'way', 'people', 'live', 'communicate', 'products', 'empower', 'people', 'express', 'live']
['snap', 'inc', 'camera', 'compani', 'believ', 'reinvent', 'camera', 'repres', 'greatest', 'opportun', 'improv', 'way', 'peopl', 'live', 'commun', 'product', 'empow', 'peopl', 'express', 'live']
['snap', 'inc', 'camera', 'company', 'believe', 'reinventing', 'camera', 'represents', 'greatest', 'opportunity', 'improve', 'way', 'people', 'live', 'communicate', 'product', 'empower', 'people', 'express', 'live']


In [102]:
df['job_description'] = df.job_description.apply(lambda text: [wnl.lemmatize(t) for t in text])
df.job_description

0      [snap, inc, camera, company, believe, reinvent...
1      [ever, wondered, happens, inside, cloud, based...
2      [roofstock, leading, marketplace, investing, s...
3      [data, engineer, responsible, processing, stru...
4      [location, tx, plano, united, state, america, ...
                             ...                        
321    [feedback, loop, feedbackloop, com, growth, st...
322    [believe, creator, ability, get, paid, value, ...
323    [embark, truck, leading, self, driving, truck,...
324    [choose, meaningful, work, fulfilling, life, m...
325    [may, mobility, may, mobility, self, driving, ...
Name: job_description, Length: 326, dtype: object

In [40]:
from nltk.util import ngrams
from collections import Counter


In [107]:
print(len(df.job_description[0]))
len(df.job_description.sum())

318


126043

In [113]:
#Create your bigrams
#ugs = nltk.unigrams(df.job_description.sum())

#compute frequency distribution for all the bigrams in the text
fdist = nltk.FreqDist(df.job_description.sum())
fdist

FreqDist({'data': 4808, 'experience': 1965, 'team': 1264, 'work': 1046, 'business': 944, 'year': 596, 'company': 585, 'technology': 576, 'skill': 568, 'tool': 567, ...})

In [121]:
sorted(fdist.items(), key=lambda item: item[1],reverse=True)[:100]

[('data', 4808),
 ('experience', 1965),
 ('team', 1264),
 ('work', 1046),
 ('business', 944),
 ('year', 596),
 ('company', 585),
 ('technology', 576),
 ('skill', 568),
 ('tool', 567),
 ('product', 546),
 ('solution', 545),
 ('development', 530),
 ('working', 525),
 ('science', 523),
 ('engineering', 521),
 ('engineer', 514),
 ('analytics', 510),
 ('python', 487),
 ('opportunity', 470),
 ('process', 451),
 ('sql', 450),
 ('learning', 446),
 ('new', 441),
 ('system', 436),
 ('design', 427),
 ('support', 413),
 ('build', 411),
 ('ability', 411),
 ('using', 408),
 ('model', 390),
 ('technical', 388),
 ('pipeline', 377),
 ('customer', 371),
 ('including', 369),
 ('environment', 361),
 ('building', 361),
 ('software', 360),
 ('analysis', 360),
 ('platform', 359),
 ('u', 356),
 ('service', 350),
 ('employee', 350),
 ('strong', 346),
 ('machine', 343),
 ('status', 340),
 ('le', 340),
 ('show', 336),
 ('time', 335),
 ('role', 330),
 ('help', 328),
 ('knowledge', 326),
 ('moreshow', 326),
 ('pro

In [122]:
#Create your bigrams
bgs = nltk.bigrams(df.job_description.sum())

#compute frequency distribution for all the bigrams in the text
fdist = nltk.FreqDist(bgs)
sorted(fdist.items(), key=lambda item: item[1],reverse=True)[:100]

[(('machine', 'learning'), 328),
 (('show', 'moreshow'), 326),
 (('moreshow', 'le'), 326),
 (('data', 'engineer'), 302),
 (('data', 'science'), 286),
 (('year', 'experience'), 242),
 (('data', 'scientist'), 229),
 (('data', 'pipeline'), 222),
 (('computer', 'science'), 186),
 (('data', 'engineering'), 165),
 (('equal', 'opportunity'), 163),
 (('experience', 'working'), 155),
 (('big', 'data'), 153),
 (('sexual', 'orientation'), 149),
 (('2', 'year'), 145),
 (('national', 'origin'), 145),
 (('data', 'set'), 136),
 (('opportunity', 'employer'), 130),
 (('experience', 'data'), 120),
 (('gender', 'identity'), 115),
 (('veteran', 'status'), 115),
 (('communication', 'skill'), 109),
 (('data', 'analytics'), 109),
 (('data', 'warehouse'), 109),
 (('3', 'year'), 107),
 (('bachelor', 'degree'), 106),
 (('e', 'g'), 101),
 (('team', 'member'), 99),
 (('data', 'source'), 96),
 (('engineering', 'team'), 93),
 (('race', 'color'), 89),
 (('data', 'model'), 87),
 (('best', 'practice'), 86),
 (('data',

In [123]:
tgs=ngrams(df.job_description.sum(),3)
fdist = nltk.FreqDist(tgs)
sorted(fdist.items(), key=lambda item: item[1],reverse=True)[:100]

[(('show', 'moreshow', 'le'), 326),
 (('equal', 'opportunity', 'employer'), 125),
 (('2', 'year', 'experience'), 78),
 (('degree', 'computer', 'science'), 77),
 (('race', 'color', 'religion'), 66),
 (('sexual', 'orientation', 'gender'), 64),
 (('without', 'regard', 'race'), 61),
 (('orientation', 'gender', 'identity'), 58),
 (('3', 'year', 'experience'), 49),
 (('bachelor', 'degree', 'computer'), 46),
 (('regard', 'race', 'color'), 45),
 (('cross', 'functional', 'team'), 44),
 (('employment', 'without', 'regard'), 41),
 (('large', 'data', 'set'), 41),
 (('gender', 'identity', 'expression'), 39),
 (('equal', 'employment', 'opportunity'), 39),
 (('5', 'year', 'experience'), 38),
 (('problem', 'solving', 'skill'), 38),
 (('medical', 'dental', 'vision'), 38),
 (('machine', 'learning', 'model'), 33),
 (('1', 'year', 'experience'), 33),
 (('color', 'religion', 'sex'), 33),
 (('year', 'experience', 'data'), 33),
 (('data', 'engineering', 'team'), 32),
 (('protected', 'veteran', 'status'), 32)